Docsp 17292 (#206)
mongodb/docs-node: source/whats-new.txt

.. contents:: On this page
   :local:
   :backlinks: none

>    :depth: 3

Reviewer: Nit: The TOC seems really long now -- just thinking ahead, it'll only be longer as more versions come out. Maybe a depth of 2 instead of 3?

Author: I agree that it could get really long, but I think it's important to make the page easily navigable. I chose 3 to allow users to quickly find and jump to different things they may be interested in.

Author: A potential solution, if needed in the future, would be to convert the What's New TOC link to a drawer and have pages for each version. That is beyond the scope of this ticket, however.

Reviewer: Agreed, this can be a future rework.

New features of the 4.0 Node.js driver release include:

TypeScript
~~~~~~~~~~

- We've migrated the driver to TypeScript. You can now harness the type
  hinting and intellisense features in editors that support it to develop

>   your MongoDB applications. You do not have to adopt TypeScript to get
>   the benefits of this work.

Reviewer 2: ```suggestion
  your MongoDB applications. You can get the benefits of this work even 
  in pure JavaScript projects!
```
Just taking a positive tone, total nit. I think we have the same lingo and I'll update too

Author: Great suggestion, thanks!

- The underlying BSON library used by this version has been migrated to
  TypeScript.

- Inline documentation has been consistenly formatted to improve display
  in editors.

- If you are a user of the community types ``@types/mongodb``, there will
  **likely be issues** adopting the types from our codebase. We could not
  achieve a one to one match in types due to the details of writing the
  codebase in TypeScript.

  We'd love to hear your TypeScript related feature requests! Please submit
  ideas on our `JIRA project here <https://jira.mongodb.org/projects/NODE>`__.

Key Changes
~~~~~~~~~~~

Node.js Version
+++++++++++++++

The minimum supported version of Node.js is now v12.9 or greater for
version 4 of the driver. Support for our 3.x branches will continue

> until summer 2022 to allow time to upgrade.

Reviewer 2: We added a note here that we should copy here that says `3.x supports back to Node.js v4`

asynchronous iteration.

``MongoClientOptions`` Interface
+++++++++++++++++++++++++++++++++

With type hinting, you should find that options passed to a ``MongoClient``
are enumerated and discoverable. We've made a large effort to process
all options in the driver to give early warnings about incompatible settings
to get your app up and running in a correct state quickly.

- ``checkServerIdentity`` is no longer checked before being passed to the
  underlying Node API. Previously, accepted values were ``false``, or
  a function. Now, the argument must be a function. Specifying a
  boolean will result in an error being thrown.

``createCollection()``
++++++++++++++++++++++

This method no longer supports a ``strict`` option, which returned
an error if the collection did not exist. To assert the existence of
a collection, use the ``listCollections()`` method instead.

.. example::

   .. code-block:: js

      const collections = (await db.listCollections({}, { nameOnly: true })
        .toArray()).map(
          ({name}) => name
        );
      if (!collections.includes(myNewCollectionName)) {
        await db.createCollection(myNewCollectionName);

>       }

Reviewer 2: Recopy this snippet, in order to replicate the old behavior we should actually throw inside the if stmt.

        });

.. note::

   ``GridFSBucket`` does not need to be closed like ``GridStore``.

File Deletion
^^^^^^^^^^^^^

.. example::

   .. code-block:: js

      // old way
      GridStore.unlink(db, name, callback);
      // new way
      bucket.delete(file_id);

Finding File Metadata
^^^^^^^^^^^^^^^^^^^^^

File metadata that used to be accessible on the GridStore instance can be
found by querying the bucket.

.. example::

   .. code-block:: js

      const fileMetaDataList: GridFSFile[] = bucket.find({}).toArray();

Unified Topology
++++++++++++++++

- We internally now only manage a ``unifiedTopology`` when you connect
  to a ``mongod``. The differences between this and previous versions
  is `detailed here <https://mongodb.github.io/node-mongodb-native/3.6/reference/unified-topology/>`__.

- You must use the new ``directConnection`` :node-api-4.0:`option </mongoclientoptions.html#directconnection>`
  to connect to unitiliazed replica set members.

Explain
+++++++

- Support has been added for fine-grained verbosity modes. You can
  learn more about each mode :manual:`here </reference/method/cursor.explain/#behavior/>`.

``instrumment()`` Removal
+++++++++++++++++++++++++

> The ``instrument()`` method has been removed.

Reviewer 2: Mention that command monitoring is the appropriate replacement for this we wrote:

- Use command monitoring: `client.on('commandStarted', (ev) => {})`

but linking to the command monitoring docs here would be great

Detailed List
~~~~~~~~~~~~~

For a detailed list of breaking changes, removals, and associated JIRA tickets,

> see the detailed `list here <https://github.com/mongodb/node-mongodb-native/blob/master/docs/CHANGES_4.0.0.md>`__.

Reviewer 2: We don't use the master branch any more in general so might wanna do an audit to make sure we don't have more links pointing there.

Should be:
`https://github.com/mongodb/node-mongodb-native/blob/4.0/docs/CHANGES_4.0.0.md`

TypeScript
~~~~~~~~~~

- We've migrated the driver to TypeScript. You can now harness the type
  hinting and intellisense features in editors that support it to develop
  your MongoDB applications. Enjoy the benefits of this work in pure JavaScript
  projects as well!

- The underlying BSON library used by this version is now migrated to
  TypeScript.

- Inline documentation is now consistenly formatted to improve display
  in editors.

- If you are a user of the community types ``@types/mongodb``, there will
  **likely be issues** adopting the types from our codebase. We could not
  achieve a one to one match in types due to the details of writing the
  codebase in TypeScript.

  We'd love to hear your TypeScript related feature requests! Please submit
  ideas on our `JIRA project here <https://jira.mongodb.org/projects/NODE>`__.

Key Changes
~~~~~~~~~~~

Node.js Version
+++++++++++++++

The minimum supported version of Node.js is now v12.9 or greater for
version 4 of the driver. Support for our 3.x branches will continue
until summer 2022 to allow time to upgrade.

.. note::

   3.x supports back to Node.js v4.

Cursor Improvements
+++++++++++++++++++

Affected classes:

> - ``AbstactCursor``

Reviewer 3: ```suggestion
- ``AbstractCursor``
```

an error if the collection did not exist. To assert the existence of
a collection, use the ``listCollections()`` method instead.

.. example::

   .. code-block:: js

      const collections = (await db.listCollections({}, { nameOnly: true })
        .toArray()).map(
          ({name}) => name
        );
      if (!collections.includes(myNewCollectionName)) {
        throw new Error(`${myNewCollectionName} doesn't exist`);
      }

``BulkWriteError`` → ``MongoBulkWriteError``
++++++++++++++++++++++++++++++++++++++++++++

``BulkWriteError`` is now renamed to ``MongoBulkWriteError``.

When running bulk operations that make writes you can encounter errors
depending on your settings. Import the new class name ``MongoBulkWriteError``
when testing for errors in bulk operations.

``DB``
++++++

``DB`` is no longer an ``EventEmitter``. Listen for events directly from your
``MongoClient`` instance.

``Collection.group()``
++++++++++++++++++++++

The ``Collection.group()`` helper, deprecated since MongoDB 3.4,
is now removed. Use the aggregation pipeline :manual:`$group </reference/operator/aggregation/group/#pipe._S_group>`
operator instead.

Authentication
++++++++++++++

> - ``gassapiServiceName`` is now removed. Use :node-api-4.0:`authMechanismProperties.SERVICE_NAME` in the URI

Reviewer 3: ```suggestion
- ``gssapiServiceName`` is now removed. Use :node-api-4.0:`authMechanismProperties.SERVICE_NAME` in the URI
```

                .on('error', console.error)
                .on('finish', () => {
                  console.log('done downloading!');
                  client.close();
                });
            });
        });

.. note::

   ``GridFSBucket`` does not need to be closed like ``GridStore``.

File Deletion
^^^^^^^^^^^^^

.. example::

   .. code-block:: js

      // old way
      GridStore.unlink(db, name, callback);
      // new way
      bucket.delete(file_id);

Finding File Metadata
^^^^^^^^^^^^^^^^^^^^^

File metadata that used to be accessible on the GridStore instance can be
found by querying the bucket.

.. example::

   .. code-block:: js

      const fileMetaDataList: GridFSFile[] = bucket.find({}).toArray();

Unified Topology
++++++++++++++++

- We internally now only manage a ``unifiedTopology`` when you connect
  to a ``mongod``. The differences between this and previous versions
  is `detailed here <https://mongodb.github.io/node-mongodb-native/3.6/reference/unified-topology/>`__.

- You must use the new ``directConnection`` :node-api-4.0:`option </mongoclientoptions.html#directconnection>`
  to connect to unitiliazed replica set members.

> 

Reviewer 3: I think it's also worth calling out that users can drop their usage of the `useUnifiedTopology` and `useNewUrlParser` options.

New features of the 4.0 Node.js driver release include:

TypeScript
~~~~~~~~~~

- We've migrated the driver to TypeScript. You can now harness the type
  hinting and intellisense features in editors that support it to develop
  your MongoDB applications. Enjoy the benefits of this work in pure JavaScript

>   projects as well!

Reviewer: Nit: I think there is something about using exclamation points in the style guide, and the last statement doesn't seem to add any value.

Author: I agree the exclam can be removed, but the last statement adds a lot of value to any user using an editor with intellisense in a JavaScript project.

- The underlying BSON library used by this version is now migrated to
  TypeScript.

- Inline documentation is now consistenly formatted to improve display
  in editors.

- If you are a user of the community types ``@types/mongodb``, there will
  **likely be issues** adopting the types from our codebase. We could not
  achieve a one to one match in types due to the details of writing the
  codebase in TypeScript.

  We'd love to hear your TypeScript related feature requests! Please submit

>   ideas on our `JIRA project here <https://jira.mongodb.org/projects/NODE>`__.

Reviewer: Issue: This currently renders as if it's part of the last bullet you have. User's may not see this unless they read the last bullet.

Key Changes
~~~~~~~~~~~

Node.js Version
+++++++++++++++

The minimum supported version of Node.js is now v12.9 or greater for
version 4 of the driver. Support for our 3.x branches will continue
until summer 2022 to allow time to upgrade.

.. note::

   3.x supports back to Node.js v4.

Cursor Improvements
+++++++++++++++++++

> Affected classes:

Reviewer: Issue: I don't know/understand how these classes were affected. I think there needs to be some wording to clarify.

- ``AbstractCursor``
- ``FindCursor``
- ``AggregationCursor``
- ``ChangeStreamCursor`` (This is the underlying cursor for ``ChangeStream``)
- ``ListCollectionsCursor``

Our Cursor implentation is now updated to make it clear what is possible

> before and after execution of an operation.

Reviewer: Suggestion: I'd want to see this part first in the section to give some context to the list of affected classes.

      stream.on("error", () => client.close())

Use ``hasNext()`` and ``next()`` for manual iteration.
Use ``for await of`` syntax or any ``Promise`` helpers for
asynchronous iteration.

``MongoClientOptions`` Interface
+++++++++++++++++++++++++++++++++

With type hinting, you should find that options passed to a ``MongoClient``
are enumerated and discoverable. We've made a large effort to process
all options in the driver to give early warnings about incompatible settings
to get your app up and running in a correct state quickly.

- ``checkServerIdentity`` is no longer checked before being passed to the
  underlying Node API. Previously, accepted values were ``false``, or
  a function. Now, the argument must be a function. Specifying a
  boolean will result in an error being thrown.

- It is longer required to specify ``useUnifiedTopology`` or ``useNewUrlParser``.

``createCollection()``
++++++++++++++++++++++

This method no longer supports a ``strict`` option, which returned
an error if the collection did not exist. To assert the existence of
a collection, use the ``listCollections()`` method instead.

.. example::

   .. code-block:: js

>       const collections = (await db.listCollections({}, { nameOnly: true })

Reviewer: Nit: Can the first line be highlighted to show where the change is?

Author: I'm not sure I understand. The change is in the underlying behavior of the method. This code block shows how to replicate the removed functionality.

Reviewer: Ah I see, I didn't realize the last sentence was describing the snippet here. Maybe ending the previous sentence with `:` would be more clear.

.. example::

   .. code-block:: language

      await client.connect();
      const filename = 'test.txt'; // whatever local file name you want
      const db = client.db();
      const bucket = new GridFSBucket(db);

      fs.createReadStream(filename)
        .pipe(bucket.openUploadStream(filename))
        .on('error', console.error)
        .on('finish', () => {
          console.log('done writing to db!');

          bucket
            .find()
            .toArray()
            .then(files => {
              console.log(files);

              bucket
                .openDownloadStreamByName(filename)
                .pipe(fs.createWriteStream('downloaded_' + filename))
                .on('error', console.error)
                .on('finish', () => {
                  console.log('done downloading!');
                  client.close();
                });
            });
        });

.. note::

   ``GridFSBucket`` does not need to be closed like ``GridStore``.

File Deletion
^^^^^^^^^^^^^

.. example::

   .. code-block:: js

      // old way
      GridStore.unlink(db, name, callback);
      // new way
      bucket.delete(file_id);

Finding File Metadata
^^^^^^^^^^^^^^^^^^^^^

> File metadata that used to be accessible on the GridStore instance can be

Reviewer: ```suggestion
File metadata that used to be accessible on the ``GridStore`` instance can be
```
